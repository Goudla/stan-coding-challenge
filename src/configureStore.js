/**
 * Store
 */
import { createMigrate, persistStore, persistReducer } from 'redux-persist';
import localForage from 'localforage';
import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { routerMiddleware } from 'connected-react-router';

// Reducers
import createRootReducer from './reducers';

// Sagas
import rootSaga from './sagas';

// Types
import type { BrowserHistory } from 'history';
import type { Store } from 'redux';

// Redux persist migrations
export const migrations = {
  '0': () => ({})
};

// Redux persist config
const config = {
  version: '0',
  key: 'primary',
  storage: localForage,
  debug: process.env.NODE_ENV !== 'production',
  whitelist: [
    'entries',
    'initialDataLoaded'
  ],
  migrate: createMigrate(migrations, { debug: false })
};

const persistedReducer = (history: BrowserHistory) =>
  persistReducer(config, createRootReducer(history));

const sagaMiddleware = createSagaMiddleware();

export default function configureStore(history: BrowserHistory) {
  // Create the store with two middlewares
  // 1. sagaMiddleware: Makes redux-sagas work
  // 2. routerMiddleware: Syncs the location/URL path to the state
  const middlewares = [sagaMiddleware, routerMiddleware(history)];

  const enhancers = [applyMiddleware(...middlewares)];

  // If Redux DevTools Extension is installed use it, otherwise use Redux compose
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

  const store: Store<any, any> = createStore(
    persistedReducer(history),
    composeEnhancers(...enhancers)
  );

  sagaMiddleware.run(rootSaga);

  let persistor = persistStore(store);

  return { persistor, store };
}
