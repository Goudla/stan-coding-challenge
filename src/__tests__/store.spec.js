import { createBrowserHistory } from 'history';

import configureStore, { migrations } from '../configureStore';

// History
const history = createBrowserHistory();

describe('configureStore', () => {
  let store;

  beforeAll(() => {
    store = configureStore(history);
  });

  describe('store', () => {
    it('should contain store', () => {
      expect(store.store).toMatchSnapshot();
    });
  });

  describe('persistor', () => {
    it('should contain persistor', () => {
      expect(store.persistor).toMatchSnapshot();
    });
  });
});

describe('redux persist migrations', () => {
  it('version 0 migrations should return an empty object', () => {
    expect(migrations['0']()).toEqual({});
  });
});
