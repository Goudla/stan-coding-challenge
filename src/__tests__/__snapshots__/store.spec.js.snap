// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configureStore persistor should contain persistor 1`] = `
Object {
  "dispatch": [Function],
  "flush": [Function],
  "getState": [Function],
  "pause": [Function],
  "persist": [Function],
  "purge": [Function],
  "replaceReducer": [Function],
  "subscribe": [Function],
  Symbol(observable): [Function],
}
`;

exports[`configureStore store should contain store 1`] = `
Object {
  "dispatch": [Function],
  "getState": [Function],
  "replaceReducer": [Function],
  "subscribe": [Function],
  Symbol(observable): [Function],
}
`;
